src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -0,0 +1,207 @@
+        else if (r.getSupplementaryAlignmentFlag) r1Supp += r
+        else if (r1.isDefined) throw new IllegalArgumentException(s"Multiple non-secondary, non-supplemental R1s for ${r.getReadName}")
+        else r1 = Some(r)
+      }
+      else {
+        if      (r.getNotPrimaryAlignmentFlag)    r2Sec += r
+        else if (r.getSupplementaryAlignmentFlag) r2Supp += r
+        else if (r2.isDefined) throw new IllegalArgumentException(s"Multiple non-secondary, non-supplemental R2s for ${r.getReadName}")
+        else r2 = Some(r)
+      }
+    }
+
+    Template(r1=r1, r2=r2, r1Supplementals=r1Supp.toList, r2Supplementals=r2Supp.toList, r1Secondaries=r1Sec.toList, r2Secondaries=r2Sec.toList)
+  }
+}
+
+/**
+  * Utility methods for working with BAMs.
+  */
+object Bams extends LazyLogging {
+  /**
+    * Builds a sorting collection for sorting SAMRecords.
+    *
+    * @param order the desired output sort order
+    * @param header the header to use to encode records during sorting
+    * @param maxInMemory the maximum number of records to keep and sort in memory
+    * @param tmpDir an optional temp directory to use for temporary sorting files
+    * @return a SortingCollection
+    */
+  def sortingCollection(order: SortOrder,
+                        header: SAMFileHeader,
+                        maxInMemory: Int = 1e6.toInt,
+                        tmpDir: Option[DirPath] = None): SortingCollection[SAMRecord] = {
+    SortingCollection.newInstance(
+      classOf[SAMRecord],
+      new BAMRecordCodec(header),
+      if (order == SortOrder.queryname) new SAMRecordQueryNameComparator else new SAMRecordCoordinateComparator,
+      maxInMemory,
+      tmpDir.map(_.toFile).orNull
+    )
+  }
+
+
+  /**
+    * Returns an iterator over the records in the given reader in such a way that all
+    * reads with the same query name are adjacent in the iterator. Does NOT guarantee
+    * a queryname sort, merely a query grouping.
+    *
+    * @param in a SamReader from which to consume records
+    * @param maxInMemory the maximum number of records to keep and sort in memory if sorting is needed
+    * @param tmpDir an optional temp directory to use for temporary sorting files if needed
+    * @return an Iterator with reads from the same query grouped together
+    */
+  def queryGroupedIterator(in: SamReader,
+                           maxInMemory: Int = 1e6.toInt,
+                           tmpDir: Option[DirPath] = None): BetterBufferedIterator[SAMRecord] = {
+    if (in.getFileHeader.getSortOrder == SortOrder.queryname || in.getFileHeader.getGroupOrder == GroupOrder.query) {
+      in.iterator().bufferBetter
+    }
+    else {
+      logger.info("Sorting into queryname order.")
+      val progress = new ProgressLogger(this.logger, "Queryname sorted")
+      val sorter = sortingCollection(SortOrder.queryname, in.getFileHeader, maxInMemory, tmpDir)
+      in.foreach { rec =>
+        sorter.add(rec)
+        progress.record(rec)
```

Comment:
I wish we had a foreach that could take a list of methods to apply.  I see us doing this a lot.